src/mumble/pluginAPI/Plugin.h

```diff
@@ -0,0 +1,242 @@
+	/// This function will also be called when freshly connecting to a server as each user on that
+	/// server needs to be "added" to the respective channel as far as the local client is concerned.
+	///
+	/// @param connection The ID of the server-connection this event is connected to
+	/// @param userID The ID of the user this event has been triggered for
+	/// @param previousChannelID The ID of the chanel the user is coming from. Negative IDs indicate that there is no previous channel (e.g. the user
+	/// 	freshly connected to the server) or the channel isn't available because of any other reason.
+	/// @param newChannelID The ID of the channel the user has entered. If the ID is negative, the new channel could not be retrieved. This means
+	/// 	that the ID is invalid.
+	PLUGIN_EXPORT void onChannelEntered(MumbleConnection_t connection, MumbleUserID_t userID, MumbleChannelID_t previousChannelID,
+			MumbleChannelID_t newChannelID);
+
+	/// Called whenever a user leaves a channel.
+	/// This function will also be called when the user disconnects from the server.
+	///
+	/// @param connection The ID of the server-connection this event is connected to
+	/// @param userID The ID of the user that left the channel
+	/// @param channelID The ID of the channel the user left. If the ID is negative, the channel could not be retrieved. This means that the ID is
+	/// 	invalid.
+	PLUGIN_EXPORT void onChannelExited(MumbleConnection_t connection, MumbleUserID_t userID, MumbleChannelID_t channelID);
+
+	/// Called when any user changes his/her talking state.
+	///
+	/// @param connection The ID of the server-connection this event is connected to
+	/// @param userID The ID of the user whose talking state has been changed
+	/// @param talkingState The new TalkingState the user has switched to.
+	PLUGIN_EXPORT void onUserTalkingStateChanged(MumbleConnection_t connection, MumbleUserID_t userID, TalkingState_t talkingState);
+
+	/// Called whenever there is audio input.
+	///
+	/// @param inputPCM A pointer to a short-array holding the pulse-code-modulation (PCM) representing the audio input.
+	/// @param inputSize The size of the input-array
+	/// @param isSpeech A boolean flag indicating whether Mumble considers the input as part of speech (instead of background noise)
+	/// @returns Whether this callback has modified the audio input-array
+	PLUGIN_EXPORT bool onAudioInput(short *inputPCM, uint32_t inputSize, bool isSpeech);
+
+	/// Called whenever there is audio output.
+	///
+	/// @param outputPCM A pointer to a short-array holding the pulse-code-modulation (PCM) representing the audio output
+	/// @param sampleCount How many samples (array-entries) there are for each channel
+	/// @param channelCount How many channels there are in the audio output
+	/// @returns Whether this callback has modified the audio output-array
+	PLUGIN_EXPORT bool onAudioOutput_short(short *outputPCM, uint32_t sampleCount, uint32_t channelCount);
```

Comment:
Where in the audio chain is this taking place? Do you have a block diagram of where this function occurs within the audio processing system?

Reply:
I intended to hook the callback somewhere into this method here:
https://github.com/mumble-voip/mumble/blob/713b92c3c233f8f38200c4a458e0196d3d39a3b0/src/mumble/AudioOutput.cpp#L355

The variable holding the actual output-data is `output`:
https://github.com/mumble-voip/mumble/blob/713b92c3c233f8f38200c4a458e0196d3d39a3b0/src/mumble/AudioOutput.cpp#L401

Probably at the end of that function after this code block:
https://github.com/mumble-voip/mumble/blob/713b92c3c233f8f38200c4a458e0196d3d39a3b0/src/mumble/AudioOutput.cpp#L565-L571

I'll try to make a block-diagram but this could take a while (I need to understand the audio flow better before I can to this).